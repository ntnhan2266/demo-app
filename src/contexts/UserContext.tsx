import React, { createContext, useState, ReactNode } from 'react';
import { PublicUser } from '@/interfaces/user';

// Context interface
interface IUserContext {
  user: PublicUser | null;
  setUser: (_user: PublicUser | null) => void;
}

// Create the context
export const UserContext = createContext<IUserContext | undefined>(undefined);

// Provider component
export const UserProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
  const [user, setUser] = useState<PublicUser | null>(null);

  const contextValue: IUserContext = {
    user,
    setUser,
  };

  return <UserContext.Provider value={contextValue}>{children}</UserContext.Provider>;
};
