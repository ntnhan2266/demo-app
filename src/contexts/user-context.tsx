import React, { createContext, useContext, useState, ReactNode } from 'react';
import { PublicUser } from '@/interfaces/user';

// Context interface
interface IUserContext {
  user: PublicUser | null;
  setUser: (user: PublicUser | null) => void;
}

// Create the context
const UserContext = createContext<IUserContext | undefined>(undefined);

// Provider component
export const UserProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
  const [user, setUser] = useState<PublicUser | null>(null);

  const contextValue: IUserContext = {
    user,
    setUser,
  };

  return <UserContext.Provider value={contextValue}>{children}</UserContext.Provider>;
};

// Custom hook to use the user context
export const useUser = () => {
  const context = useContext(UserContext);

  if (!context) {
    throw new Error('useUser must be used within a UserProvider');
  }

  return context;
};
