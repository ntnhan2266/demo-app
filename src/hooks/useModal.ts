import { useState, useCallback } from 'react';

interface IModalHook {
  isOpen: boolean;
  openModal: () => void;
  closeModal: () => void;
}

const useModal = (): IModalHook => {
  const [isOpen, setIsOpen] = useState(false);

  const openModal = useCallback(() => {
    setIsOpen(true);
  }, []);

  const closeModal = useCallback(() => {
    setIsOpen(false);
  }, []);

  return { isOpen, openModal, closeModal };
};

export default useModal;
