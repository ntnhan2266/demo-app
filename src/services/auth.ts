import { IUser } from '@/interfaces/user';
import { loginUser as mockLoginUser, registerUser as mockRegisterUser } from '@/utils/mock-user-database';

export const registerUser = (user: IUser): boolean => {
  // Just call mock function to simulate api call
  return mockRegisterUser(user);
};

export const loginUser = (emailOrPhone: string, password: string): IUser | null => {
  const matchedUser = mockLoginUser(emailOrPhone, password);

  if (matchedUser) {
    // Return user info upon successful login
    return matchedUser;
  }

  return null; // Return null for unsuccessful login
};
