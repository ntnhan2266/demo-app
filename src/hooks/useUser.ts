import { useContext } from "react";
import { UserContext } from "@/contexts/UserContext";

// Custom hook to use the user context
export const useUser = () => {
  const context = useContext(UserContext);

  if (!context) {
    throw new Error('useUser must be used within a UserProvider');
  }

  return context;
};
