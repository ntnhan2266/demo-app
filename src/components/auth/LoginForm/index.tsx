import React, { useState } from 'react';
import { useForm } from 'react-hook-form';
import { useNavigate } from 'react-router-dom';
import { loginUser } from '@/services/auth';
import Input from '@/components/common/Input';
import Button from '@/components/common/Button';
import { ROUTE_PATH } from '@/constants/route-path';
import { isValidContact } from '@/utils/validation';
import { useUser } from '@/hooks/useUser';

interface ILoginFormInputs {
  emailOrPhone: string;
  password: string;
}

const LoginForm: React.FC = (): React.ReactElement => {
  const navigate = useNavigate();
  const {
    register,
    handleSubmit,
    reset,
    formState: { errors },
  } = useForm<ILoginFormInputs>();
  const { setUser } = useUser();
  const [successMessage, setSuccessMessage] = useState<string | null>(null);
  const [errorMessage, setErrorMessage] = useState<string | null>(null);

  const onSubmit = (values: ILoginFormInputs): void => {
    const user = loginUser(values.emailOrPhone, values.password);

    if (!user) {
      setErrorMessage('Login failed. Invalid credentials.');
      return;
    }

    // For simplicity, just display a success message
    setSuccessMessage('Login successful!');

    // Update authenticated user data into context
    setUser(user);

    // Reset the form
    // Note: Use reset() if you need to clear the form fields
    reset();

    // Navigate to the dashboard or any other page after login
    navigate(ROUTE_PATH.DASHBOARD_UPDATE_INFO);
  };

  return (
    <div data-testid='login-form' className='max-w-md mx-auto mt-8 p-6 bg-white rounded-md shadow'>
      <h2 className='text-2xl font-bold mb-4 uppercase'>Login to your account</h2>
      <form onSubmit={handleSubmit(onSubmit)}>
        <Input
          {...register('emailOrPhone', {
            required: 'Email or phone is required.',
            validate: (val: string) => isValidContact(val) || 'Invalid email or phone.',
          })}
          errorMessage={errors.emailOrPhone?.message}
          type='text'
          label='Email or phone'
          placeholder='Enter your email or phone number'
          wrapperClass='mb-4'
          data-testid='email-or-phone-input'
          inputClasses='w-full md:w-full'
          id='email'
          required
        />
        <Input
          {...register('password', {
            required: 'Password is required.',
          })}
          errorMessage={errors.password?.message}
          type='password'
          label='Password'
          placeholder='Enter your password'
          wrapperClass='mb-4'
          data-testid='password-input'
          inputClasses='w-full md:w-full'
          id='password'
          required
        />
        <Button data-testid='submit-button' label='Login' type='submit' fullWidth />
      </form>
      {successMessage && (
        <div data-testid className='mt-4 p-3 bg-green-200 text-green-800 rounded-md'>
          {successMessage}
        </div>
      )}
      {errorMessage && (
        <div data-testid='error-message' className='mt-4 p-3 bg-red-200 text-red-800 rounded-md'>
          {errorMessage}
        </div>
      )}
    </div>
  );
};

export default LoginForm;
