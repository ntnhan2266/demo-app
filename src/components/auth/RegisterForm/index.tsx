import React, { useState } from 'react';
import { useForm } from 'react-hook-form';
import Button from '@/components/common/Button';
import Input from '@/components/common/Input';
import { isValidContact } from '@/utils/validation';
import { registerUser } from '@/services/auth';
import ErrorModal from '@/components/common/ErrorModal';
import SuccessModal from '@/components/common/SuccessModal';

interface IFormInputs {
  firstName: string;
  lastName: string;
  emailOrPhone: string;
  password: string;
  confirmPassword: string;
}

const RegisterForm: React.FC = (): React.ReactElement => {
  const {
    register,
    handleSubmit,
    reset,
    watch,
    setError,
    formState: { errors },
  } = useForm<IFormInputs>({
    defaultValues: {
      firstName: '',
      lastName: '',
      emailOrPhone: '',
      password: '',
      confirmPassword: '',
    },
  });
  const [isErrorModalOpen, setIsErrorModalOpen] = useState(false);
  const [isSuccessModalOpen, setIsSuccessModalOpen] = useState(false);

  const closeErrorModal = (): void => {
    // Close the error modal
    setIsErrorModalOpen(false);
  };

  const closeSuccessModal = (): void => {
    // Close the success modal
    setIsSuccessModalOpen(false);
  };

  const onSubmit = (values: IFormInputs): void => {
    const result = registerUser(values);
    if (!result) {
      // Registration failed
      setIsErrorModalOpen(true);
      setError('emailOrPhone', {
        message: 'This email or phone is already in use.',
      });
      return;
    }

    // Registration successful
    setIsSuccessModalOpen(true);

    // Reset the form
    reset();
  };

  return (
    <div data-testid='register-form' className='max-w-md mx-auto mt-8 p-6 bg-white rounded-md shadow'>
      <h2 className='text-2xl font-bold mb-4 uppercase'>Create your account</h2>
      <form onSubmit={handleSubmit(onSubmit)}>
        <Input
          {...register('firstName', { required: 'First name is required.' })}
          errorMessage={errors.firstName?.message}
          type='text'
          label='First name'
          placeholder='Enter your first name'
          wrapperClass='mb-4'
          data-testid='first-name-input'
          inputClasses='w-full md:w-full'
          required
        />
        <Input
          {...register('lastName', { required: 'Last name is required.' })}
          errorMessage={errors.lastName?.message}
          type='text'
          label='Last name'
          placeholder='Enter your last name'
          wrapperClass='mb-4'
          data-testid='last-name-input'
          inputClasses='w-full md:w-full'
          required
        />
        <Input
          {...register('emailOrPhone', {
            required: 'Email or phone is required.',
            validate: (val: string) => isValidContact(val) || 'Invalid email or phone.',
          })}
          errorMessage={errors.emailOrPhone?.message}
          type='text'
          name='emailOrPhone'
          label='Email or phone'
          placeholder='Enter either your email or phone number'
          wrapperClass='mb-4'
          data-testid='email-or-phone-input'
          inputClasses='w-full md:w-full'
          required
        />
        <Input
          {...register('password', {
            required: 'Password is required.',
            pattern: {
              value: /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[!@#$%^&*(),.?":{}|<>])[A-Za-z\d!@#$%^&*(),.?":{}|<>]{12,}$/,
              message:
                'Password must be strong (include uppercase, lowercase, number, special character, and at least 12 characters).',
            },
          })}
          errorMessage={errors.password?.message}
          type='password'
          label='Password'
          placeholder='Enter your password'
          wrapperClass='mb-4'
          data-testid='password-input'
          inputClasses='w-full md:w-full'
          required
        />
        <Input
          {...register('confirmPassword', {
            required: 'Confirm password is required.',
            validate: (val: string) => {
              if (watch('password') != val) {
                return 'Your passwords do not match.';
              }
            },
          })}
          errorMessage={errors.confirmPassword?.message}
          type='password'
          label='Confirm password'
          placeholder='Confirm your password'
          wrapperClass='mb-4'
          data-testid='confirm-password-input'
          inputClasses='w-full md:w-full'
          required
        />
        <Button data-testid='submit-button' label='Register' type='submit' fullWidth />
      </form>
      <ErrorModal
        title='Failure!'
        content='There was an error during the registration process. Please try again.'
        isOpen={isErrorModalOpen}
        onClose={closeErrorModal}
      />
      <SuccessModal
        title='Success!'
        content='Your account has been successfully registered.'
        isOpen={isSuccessModalOpen}
        onClose={closeSuccessModal}
      />
    </div>
  );
};

export default RegisterForm;
