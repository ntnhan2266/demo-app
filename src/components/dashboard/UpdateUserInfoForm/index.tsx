import React, { useState } from 'react';
import { useForm } from 'react-hook-form';
import Button from '@/components/common/Button';
import Input from '@/components/common/Input';
import { updateUserInfo } from '@/services/auth';
import ErrorModal from '@/components/common/ErrorModal';
import SuccessModal from '@/components/common/SuccessModal';
import { useUser } from '@/hooks/useUser';

interface IFormInputs {
  firstName: string;
  lastName: string;
}

const UpdateInfoForm: React.FC = (): React.ReactElement => {
  const { user, setUser } = useUser();
  const {
    register,
    handleSubmit,
    reset,
    formState: { errors },
  } = useForm<IFormInputs>({
    defaultValues: {
      firstName: '',
      lastName: '',
    },
  });
  const [errorMessage, setErrorMessage] = useState<string | null>(null);
  const [isErrorModalOpen, setIsErrorModalOpen] = useState(false);
  const [isSuccessModalOpen, setIsSuccessModalOpen] = useState(false);

  const closeErrorModal = (): void => {
    setIsErrorModalOpen(false);
  };

  const closeSuccessModal = (): void => {
    setIsSuccessModalOpen(false);
  };

  const onSubmit = (values: IFormInputs): void => {
    setErrorMessage(null);

    if (!user) {
      setErrorMessage('User not found');
      setIsErrorModalOpen(true);
      return;
    }

    if (!values.lastName && !values.firstName) {
      return;
    }

    const result = updateUserInfo(user.emailOrPhone, values);
    if (!result) {
      setIsErrorModalOpen(true);
      return;
    }

    setUser(result);
    setIsSuccessModalOpen(true);
    reset(); // Reset the form on successful update
  };

  return (
    <div data-testid='update-user-info-form'>
      <h2 className='text-2xl font-bold mb-4'>Update your information</h2>
      <form onSubmit={handleSubmit(onSubmit)}>
        <Input disabled type='text' label='Email or phone' wrapperClass='mb-4' value={user?.emailOrPhone} />
        <Input
          {...register('firstName')}
          errorMessage={errors.firstName?.message}
          type='text'
          label='First name'
          placeholder='Enter your first name'
          wrapperClass='mb-4'
          data-testid='first-name-input'
        />
        <Input
          {...register('lastName')}
          errorMessage={errors.lastName?.message}
          type='text'
          label='Last name'
          placeholder='Enter your last name'
          wrapperClass='mb-4'
          data-testid='last-name-input'
        />
        <Button data-testid='submit-button' label='Update Info' type='submit' className='w-96' />
      </form>
      <ErrorModal
        title='Failure!'
        content={errorMessage || 'There was an error during the update process. Please try again.'}
        isOpen={isErrorModalOpen}
        onClose={closeErrorModal}
      />
      <SuccessModal
        title='Success!'
        content='Your account has been successfully updated.'
        isOpen={isSuccessModalOpen}
        onClose={closeSuccessModal}
      />
    </div>
  );
};

export default UpdateInfoForm;
