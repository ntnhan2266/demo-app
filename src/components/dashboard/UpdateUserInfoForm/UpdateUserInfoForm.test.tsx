import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import UpdateUserInfoForm from '@/components/dashboard/UpdateUserInfoForm';
import { updateUserInfo } from '@/services/auth';
import renderer from 'react-test-renderer';
import { IUserContext, UserContext } from '@/contexts/UserContext';
import { useUser } from '@/hooks/useUser';

jest.mock('@/hooks/useUser');
jest.mock('@/services/auth');

beforeEach(() => {
  jest.clearAllMocks();
});

const userData = {
  firstName: 'John',
  lastName: 'Doe',
  emailOrPhone: 'john.doe@example.com',
  password: '$2a$10$Bb86lbi8ioa/UgwOVBjqTu0SzGEMu8lruBeoQZ4LZiBKUyNg1QTcG',
};

const contextValue: IUserContext = {
  user: userData,
  setUser: jest.fn(), // Mock setUser function
};

describe('UpdateUserInfoForm component', () => {
  const mockUseUser = useUser as jest.Mock;
  const mockUpdateUserInfo = updateUserInfo as jest.Mock;
  const mockSetUser = jest.fn();

  beforeEach(() => {
    mockUseUser.mockReturnValue({ user: { emailOrPhone: 'test@example.com' }, setUser: jest.fn() });
    render(
      <UserContext.Provider value={contextValue}>
        <UpdateUserInfoForm />
      </UserContext.Provider>,
    );
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('should match the snapshot', () => {
    const tree = renderer.create(<UpdateUserInfoForm />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('renders the form with the correct inputs', () => {
    const firstNameInput = screen.getByTestId('first-name-input');
    const lastNameInput = screen.getByTestId('last-name-input');
    const updateButton = screen.getByTestId('submit-button');

    expect(firstNameInput).toBeInTheDocument();
    expect(lastNameInput).toBeInTheDocument();
    expect(updateButton).toBeInTheDocument();
  });

  test('calls updateUserInfo function when form is submitted with valid data', async () => {
    mockUseUser.mockReturnValue({ user: { emailOrPhone: 'test@example.com' }, setUser: mockSetUser });
    mockUpdateUserInfo.mockReturnValue({ emailOrPhone: 'test@example.com', firstName: 'John', lastName: 'Doe' });

    // Fill in the form
    const firstNameInput = screen.getByTestId('first-name-input');
    const lastNameInput = screen.getByTestId('last-name-input');
    const updateButton = screen.getByTestId('submit-button');

    fireEvent.input(firstNameInput, { target: { value: 'John' } });
    fireEvent.input(lastNameInput, { target: { value: 'Doe' } });
    fireEvent.click(updateButton);

    // Wait for the success modal to appear
    await waitFor(() => {
      expect(mockUpdateUserInfo).toHaveBeenCalledWith('test@example.com', { firstName: 'John', lastName: 'Doe' });
      expect(screen.getByText('Your account has been successfully updated.')).toBeInTheDocument();
    });
  });
});
