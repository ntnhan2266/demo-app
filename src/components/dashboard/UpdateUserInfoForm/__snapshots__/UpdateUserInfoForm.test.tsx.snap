// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateUserInfoForm component should match the snapshot 1`] = `
<div
  data-testid="update-user-info-form"
>
  <h2
    className="text-2xl font-bold mb-8 uppercase"
  >
    Update your information
  </h2>
  <form
    onSubmit={[Function]}
  >
    <div
      className="mb-4"
    >
      <label
        className="block w-full mb-1 text-xs font-medium text-gray-500 transition-all duration-200 ease-in-out "
      >
        Email or phone
      </label>
      <input
        className="border rounded-md px-3 py-2 focus:outline-none transition duration-300 w-full md:w-96 max-w-full  bg-gray-200 cursor-not-allowed"
        disabled={true}
        placeholder=""
        type="text"
        value="test@example.com"
      />
    </div>
    <div
      className="mb-4"
    >
      <label
        className="block w-full mb-1 text-xs font-medium text-gray-500 transition-all duration-200 ease-in-out "
        htmlFor="firstName"
      >
        First name
      </label>
      <input
        className="border rounded-md px-3 py-2 focus:outline-none transition duration-300 w-full md:w-96 max-w-full  border-gray-400"
        data-testid="first-name-input"
        disabled={false}
        name="firstName"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Enter your first name"
        type="text"
      />
    </div>
    <div
      className="mb-4"
    >
      <label
        className="block w-full mb-1 text-xs font-medium text-gray-500 transition-all duration-200 ease-in-out "
        htmlFor="lastName"
      >
        Last name
      </label>
      <input
        className="border rounded-md px-3 py-2 focus:outline-none transition duration-300 w-full md:w-96 max-w-full  border-gray-400"
        data-testid="last-name-input"
        disabled={false}
        name="lastName"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Enter your last name"
        type="text"
      />
    </div>
    <button
      className="rounded-md focus:outline-none w-96 text-white text-base px-4 py-2 bg-indigo-500 hover:bg-indigo-600"
      data-testid="submit-button"
      disabled={false}
      type="submit"
    >
      Update Info
    </button>
  </form>
</div>
`;
